### PR TITLE
Implement TypedArray#forEach and TypedArray#filter

### PR DESCRIPTION
**Pull request by jtenner**
This pull request aims to implement the `forEach` and `filter` functions on the `TypedArray` prototype.

Thanks to @MaxGraey for putting up with my annoying questions and bugs.

### CODE REVIEW

File: std/assembly/internal/typedarray.ts [hunk 1]
```diff
@@ -231,3 +232,47 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+    callbackfn(LOAD<T>(buffer, i, byteOffset), i, array);
+  }
+}
+
+@inline
+export function FILTER<TArray extends TypedArray<T>, T>(
+  array: TArray,
+  callbackfn: (value: T, index: i32, array: TArray) => bool,
+): TArray {
+  var backingBuffer = new ArrayBuffer(sizeof<T>());
```
Comment (dcodeIO) [inline std/assembly/internal/typedarray.ts]: Should probably use `allocateUnsafe` directly. Safes a few checks, and doesn't zero memory.

File: std/assembly/internal/typedarray.ts [hunk 2]
```diff
@@ -231,3 +232,47 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+  var length = array.length;
+  var byteOffset = array.byteOffset;
+  var backByteLength = backingBuffer.byteLength;
+  var value: T;
+  for (let i = 0; i < length; i++) {
+    value = LOAD<T>(buffer, i, byteOffset);
+    if (callbackfn(value, i, array)) {
+      backByteLength = backingBuffer.byteLength;
+      if (backByteLength < (count + 1) << alignof<T>()) {
+        backingBuffer = reallocateUnsafe(backingBuffer, backByteLength << 1);
```
Comment (dcodeIO) [inline std/assembly/internal/typedarray.ts]: Not related to your PR directly: One thing that bothers me about reallocateUnsafe vs allocateUnsafe is that the former zeroes memory whenever newSize > oldSize, while the latter doesn't bother with zeroing anything. Somewhat inconsistent and thus likely to change, in turn making it necessary to review any code that uses it.

File: std/assembly/internal/typedarray.ts (hunk 1 above)
Comment (jtenner) [reply std/assembly/internal/typedarray.ts]: Okay. For now it seems wise to call `allocateUnsafe()` directly.

File: std/assembly/internal/typedarray.ts (hunk 2 above)
Comment (jtenner) [reply std/assembly/internal/typedarray.ts]:
```ts
```ts
export function growUnsafe(buffer: ArrayBuffer, newByteLength: i32): ArrayBuffer {
  var oldByteLength = buffer.byteLength;
  assert(oldByteLength < newByteLength && newByteLength < MAX_BLENGTH);
  if (newByteLength <= <i32>(computeSize(oldByteLength) - HEADER_SIZE)) {
    store<i32>(changetype<usize>(buffer), newByteLength, offsetof<ArrayBuffer>("byteLength"));
  } else {
    let newBuffer = allocateUnsafe(newByteLength);
    memory.copy(
      changetype<usize>(newBuffer) + HEADER_SIZE,
      changetype<usize>(buffer) + HEADER_SIZE,
      <usize>oldByteLength
    );
    if (!isManaged<ArrayBuffer>()) {
      memory.free(changetype<usize>(buffer));
    }
    buffer = newBuffer;
  }
  return buffer;
}
```

**Comment by jtenner** [comment]
Why is travis suddenly failing? I didn't modify ArrayBuffer or TypedArray?

File: std/assembly/internal/typedarray.ts (hunk 2 above)
Comment (MaxGraey) [reply std/assembly/internal/typedarray.ts]: Sorry for relay. But `growUnsafe` is unnecessary because we already have `reallocateUnsafe` which do the same

Comment (jtenner) [reply std/assembly/internal/typedarray.ts]: Alright. I switch my strategy to this.

File: std/assembly/internal/arraybuffer.ts [hunk 3]
```diff
@@ -34,7 +34,7 @@ export function allocateUnsafe(byteLength: i32): ArrayBuffer {
   return changetype<ArrayBuffer>(buffer);
 }
 
-export function reallocateUnsafe(buffer: ArrayBuffer, newByteLength: i32): ArrayBuffer {
+export function reallocateUnsafe(buffer: ArrayBuffer, newByteLength: i32, zeroFill: bool = false): ArrayBuffer {
```
Comment (MaxGraey) [inline std/assembly/internal/arraybuffer.ts]: Sorry, if we don't want break existing behaviours `zeroFill` should be `true` by default

Comment (jtenner) [reply std/assembly/internal/arraybuffer.ts]: I was following your example as you presented it. I'll make the fix now.

Comment (MaxGraey) [reply std/assembly/internal/arraybuffer.ts]: Yeah, my bad

**Comment by jtenner** [comment]
This is ready for another review.

File: std/assembly/internal/typedarray.ts [hunk 4]
```diff
@@ -231,3 +232,52 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+      ++count;
+    }
+  }
+
+  // reallocateUnsafe calls this to reduce ArrayBuffer size
+  store<i32>(changetype<usize>(backingBuffer) + offsetof<ArrayBuffer>("byteLength"), count << alignof<T>());
+
+  // instantiate the result
+  var result = instantiate<TArray>(0);
+  store<ArrayBuffer>(changetype<usize>(result) + offsetof<TArray>("buffer"), backingBuffer);
```
Comment (MaxGraey) [inline std/assembly/internal/typedarray.ts]: Could you try move `offsetof<TArray>("buffer")` and `offsetof<TArray>("byteLength")` which actually compile time consts to last `offset` param like:
```ts
store<ArrayBuffer>(changetype<usize>(result), backingBuffer, offsetof<TArray>("buffer"));
```

File: std/assembly/internal/typedarray.ts [hunk 5]
```diff
@@ -231,3 +232,52 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+  store<i32>(changetype<usize>(result) + offsetof<TArray>("byteLength"), backingBuffer.byteLength);
```
Comment (MaxGraey) [inline std/assembly/internal/typedarray.ts]: Same for `offsetof<TArray>("byteLength")`

File: std/assembly/internal/typedarray.ts [hunk 6]
```diff
@@ -231,3 +232,52 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+  var buffer = array.buffer;
+  var length = array.length;
+  var byteOffset = array.byteOffset;
+  var backByteLength = backingBuffer.byteLength;
+  var value: T;
+  for (let i = 0; i < length; i++) {
+    value = LOAD<T>(buffer, i, byteOffset);
+    if (callbackfn(value, i, array)) {
+      backByteLength = backingBuffer.byteLength;
+      if (backByteLength < (count + 1) << alignof<T>()) {
```
Comment (MaxGraey) [inline std/assembly/internal/typedarray.ts]: Not sure about this check. I guess better for this in this way:
```ts
if ((backByteLength >> alignof<T>()) < count + 1)
```
cc @dcodeIO

File: std/assembly/internal/typedarray.ts (hunk 4 above)
Comment (jtenner) [reply std/assembly/internal/typedarray.ts]: Sounds great! I had no idea this was possible.

**Comment by dcodeIO** [comment]
My issues here aren't so much with this PR specifically, but that I am unhappy with how this has to be implemented currently. We'd really need better / more consistent helpers here. Hence hasitating to merge even more code that will most likely explode when GC is turned on.

**Comment by jtenner** [comment]
@dcodeIO I will close this request and then implement `forEach` alone on it's own branch if that's okay.

**Comment by dcodeIO** [comment]
Well, not saying that this needs to be closed. Just somewhat hard for me (maybe it's actually just me) to wrap my head around every single scenario that could go wrong, with the possibility on the horizon that I'll just go and throw away some code to make this whole process easier and more reliable. That requires me to dive back into all the GC internals, though, which would be a distraction from trying to get other stuff fixed that I'm currently looking into.

**Comment by jtenner** [comment]
Yeah let's close this :)

**Pull request closed without merging**